pilot/pkg/model/policyattachment.go

```diff
@@ -21,33 +21,62 @@ import (
 	"istio.io/istio/pkg/config/constants"
 	"istio.io/istio/pkg/config/labels"
 	"istio.io/istio/pkg/config/schema/gvk"
+	"istio.io/istio/pkg/slices"
+	"istio.io/istio/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/types"
 )
 
-type policyTargetGetter interface {
+// this can be any type from istio/api that uses these types of selectors
+type TargetablePolicy interface {
 	GetTargetRef() *v1beta1.PolicyTargetReference
 	GetSelector() *v1beta1.WorkloadSelector
 }
 
-type WorkloadSelectionOpts struct {
-	RootNamespace  string
+// WorkloadPolicyMatcher performs policy selection either using targetRef or label selectors.
+// Label selection uses the workload labels.
+// TargetRef selection uses either the workload's namespace + the gateway name based on labels,
+// or the Services the workload is a part of.
+type WorkloadPolicyMatcher struct {
 	Namespace      string
 	WorkloadLabels labels.Instance
 	IsWaypoint     bool
+	// TODO can this just be sets.String since the services must be same-namespace?
+	Services sets.Set[types.NamespacedName]
 }
 
-type policyMatch string
+func PolicyMatcherFor(workloadNamespace string, labels labels.Instance, isWaypoint bool) WorkloadPolicyMatcher {
+	return WorkloadPolicyMatcher{
+		Namespace:      workloadNamespace,
+		WorkloadLabels: labels,
+		IsWaypoint:     isWaypoint,
+		// TODO check if any callers need services (i think it's just mTLS checker)
+		Services: map[types.NamespacedName]struct{}{},
+	}
+}
 
-const (
-	// policyMatchSelector is the default behavior. If the workload matches the policy's selector, the policy is applied
-	policyMatchSelector policyMatch = "selector"
-	// policyMatchDirect is used when the policy has a targetRef, and the workload matches the targetRef.
-	// Note that the actual targetRef matching is done within `getPolicyMatcher`
-	policyMatchDirect policyMatch = "direct"
-	// policyMatchIgnore indicates that there is no match between the workload and the policy, and the policy should be ignored
-	policyMatchIgnore policyMatch = "ignore"
-)
+func PolicyMatcherForProxy(proxy *Proxy, push *PushContext) WorkloadPolicyMatcher {
+	var proxyServices  sets.Set[types.NamespacedName]
+	if proxy.IsWaypointProxy() {
+		if push != nil {
+			key := WaypointKeyForProxy(proxy)
+			proxyServices = sets.New(slices.Map(push.ServicesForWaypoint(key), ServiceInfo.NamespacedName)...)
```

Comment:
I think this isn't quite right. Say we have 2 services. We should have

```
listener: main_internal
filterchains:
 - name: service 1
   rbac: rbac for service 1
 - name: service 2
   rbac: rbac for service 2
```

But this implements:


```
listener: main_internal
filterchains:
 - name: service 1
   rbac: rbac for service 1+2
 - name: service 2
   rbac: rbac for service 2+2
```

Reply:
I changed this, but it resulted in changing a ton of the listener generation code :/